Vb 2 get volunteers by filter

serghei-rubailo opened the pull request:
### What does it fix?

Closes #85 

1. Add filter by fields phone, zone, role in get_volunteers_by_filters method
2. Add test volunteers creation in seeds.py
3. Remove password filed from register_volunteer, because Volunteer model doesn't have password field
4. Add clean_data method to Volunteer model, without it we can't get volunteer


### How has it been tested?

1. http://localhost:5000/api/volunteer/filters/1/20 - checked that all volunteers are shown
2. http://localhost:5000/api/volunteer/filters/1/2?phone=25&role=delivery&zone=Botanica - checked that volunteer is found according filters
3. http://localhost:5000/api/volunteer/filters/1/2?phone=66 - checked that volunteer can't be found if there are no volunteer with such phone (role, zone)

P.S. It looks like filter worked before  my changes and we can search via phone, zone, role parameters
Please reply if filter should search for, for example, part of the phone number.

alexciornii commented on backend/endpoints/volunteer.py:
```diff
@@ -181,7 +180,7 @@ def get_volunteers_by_filters(filters, pages=0, per_page=10000):
             flt = {}
             toBool = {"true": True, "false": False}
-            caseS = ["first_name", "last_name"]
+            caseS = ["first_name", "last_name", "phone", "zone", "role"]
```
Sorry, there is not enough information in the job description. Volunteers should be filtered either by first name, last name, or by phone. At the same time, they can be additionally filtered by zone and role. Here is  the [design](https://www.figma.com/proto/oCN3NSECKQnS4PWhZnWGCT/Dashboard_Voluntar.md_V2.0?node-id=585%3A4930&viewport=402%2C967%2C0.4029495418071747&scaling=scale-down-width)

alexciornii commented on backend/models/volunteer_model.py:
```diff
@@ -68,3 +68,12 @@ class Volunteer(Document):
     aggreed_terms = BooleanField(default=False)
     april1 = BooleanField(default=False)
+
+    def clean_data(self) -> dict:
```
What is it added for?

serghei-rubailo replied:
It was added to return JSON result in existing format
 {
  "count": 3, 
  "list": [
    {
      "_id": "5f712ab2feee02c4a1e7a005" ....

I can remove clean_data method and use something like this to return JSON
`volunteers = [**json.loads(v.to_json())** for v in obj.order_by("-created_at").skip(offset).limit(item_per_age)]
return jsonify({"list": volunteers, "count": obj.count()})`

It will return almost the same JSON structure, but difference will be in next fields:
`"_id": {
        "$oid": "5f712ab2feee02c4a1e7a005"
      }`
`"created_at": {
        "$date": 1601252018190
      }`

If it will not be a problem I'll remove clean_data method from volunteer_model and use **json.loads(v.to_json())**

serghei-rubailo commented:
Hi,
Please review changes in pull request.
Additional points to first reply in pull request:

1. As I understand according [prototype ](https://www.figma.com/proto/oCN3NSECKQnS4PWhZnWGCT/Dashboard_Voluntar.md_V2.0?node-id=585%3A4930&viewport=402%2C967%2C0.4029495418071747&scaling=scale-down-width) one search field will be used for searching Volunteer First Name/Last Name/Phone.
That means api will receive only one field in GET request for First Name/Last Name/Phone, so I've changed it to _query_

According to my idea script will search for first word from the _query_ in first_name/last_name/phone database fields and after will search for second (third and etc.) word in QuerySet from search for first word.

2. For additional search there are three fields "zone", "role", "status".
"status" was added out of scope of this task, but as see in prototype Volunteers will be search by a status as well and "is_active" field will not be used, so I didn't add "is_active" to search fields.

3. If unknown key will be sent in request next error will be returned: key + " key can't be found"

4. Volunteer phone field was changed to String type according to Class Diagram. Phone can consist next symbols: ( ) + -.
If one search field will be used for First Name/Last Name/Phone, then it is better to save phone as String. 


**How has it been tested?**

1. http://localhost:5000/api/volunteer/filters/1/50?query=S - all Volunteers with "s" in first_name/last_name/phone are returned
2. http://localhost:5000/api/volunteer/filters/1/50?query=Serghei - Volunteers with "serghei " in first_name/last_name/phone are returned
3. http://localhost:5000/api/volunteer/filters/1/50?query=Serghei%20Volkov - Volunteer Serghei Volkov is returned
4. http://localhost:5000/api/volunteer/filters/1/50?query=Serghei&zone=Botanica - all Sergheis from Botanica are returned
5. http://localhost:5000/api/volunteer/filters/1/50?query=Serghei&zone=Botanica&role=copilot - nothing is return because we don't have in test dataset copilots Serghei from Botanica
6. http://localhost:5000/api/volunteer/filters/1/50?query=Serghei&zone=Botanica&role=delivery - delivery Serghei from Botanica is returned

alexciornii commented on backend/endpoints/volunteer.py:
```diff
@@ -180,22 +182,50 @@ def get_volunteers_by_filters(filters, pages=0, per_page=10000):
+        search_word = search_words_list[0]
+
+        if search_result is None:
+            search_result = Volunteer.objects(Q(first_name__icontains=search_word) | Q(last_name__icontains=search_word) |
```
Please use operator `istartswith`

serghei-rubailo replied:
Done, istartswith is used for search

alexciornii commented on backend/endpoints/volunteer.py:
```diff
@@ -180,22 +182,50 @@ def get_volunteers_by_filters(filters, pages=0, per_page=10000):
+            search_result = Volunteer.objects(Q(first_name__icontains=search_word) | Q(last_name__icontains=search_word) |
+                                    Q(phone__icontains=search_word))
+        else:
+            search_result = search_result.filter(Q(first_name__icontains=search_word) | Q(last_name__icontains=search_word) |
```
Please use operator `istartswith`

serghei-rubailo replied:
Done, istartswith is used for search

alexciornii commented on backend/seeds.py:
```diff
@@ -35,6 +62,24 @@ def seed_db_command():
+                "phone": volunteer.phone,
+                "zone": volunteer.zone,
+                "address": volunteer.address,
+                "zone_address": volunteer.zone_address,
```
Change this field as not require and this row can be removed

serghei-rubailo replied:
zone_address changed to not required

alexciornii commented on backend/seeds.py:
```diff
@@ -13,17 +16,41 @@ class SeedUser(NamedTuple):
+    zone: Zone
+    address: str
+    role: VolunteerRole
+    zone_address: str
```
can be removed

serghei-rubailo replied:
Done, zone_addres is removed

alexciornii commented on backend/endpoints/volunteer.py:
```diff
@@ -180,22 +182,50 @@ def get_volunteers_by_filters(filters, pages=0, per_page=10000):
         return jsonify({"error": str(error)}), 400
 
 
+def volunteer_keywords_search(search_words_list, search_result=None):
```
Let's generalize this method and move it to another file so other models can use it.

serghei-rubailo replied:
Method is generalized, renamed to model_keywords_search and moved to utils/search.py

serghei-rubailo commented:
Hello! Thank you for previous review.
Please check next changes:

1. We use __istartswith for search in database fields
2. Check for empty search query was added, ?query= returned error, now it returns all results because no keywords are provided
3. zone_address is not required Volunteer field
4. New files was created /utils/search.py, it contains model_keywords_search() method
model_keywords_search method should receive class where to search, which fields should be searched and keywords to search.
Now search fields can be changed dynamically when model_keywords_search is called.

johnsmithm commented on backend/models/volunteer_model.py:
```diff
@@ -22,25 +22,25 @@ class Tags(Document):
     first_name = StringField(max_length=500, required=True)
     last_name = StringField(max_length=500, required=True)
-    phone = IntField(min_value=16, max_value=120)
+    phone = StringField(max_length=255)
```
max_length for a phone number is 8, and maybe add a regex check to be only digits?

serghei-rubailo replied:
Yes, you are right, regex will be added to accept only digits.
Ok, I will set max_length = 8 numbers, but we will not be able to enter mobile phone number with leading zero or full number with country code + city code
Ex.: 
069000000 - 9 numbers
37322000000 - country code + city code + landline phone number
If it will not be a problem, then OK )

alexciornii replied:
@johnsmithm If you don't have any questions on this, please resolve it

serghei-rubailo replied:
Ok, thank you!

johnsmithm commented:
also update the filter section from the swagger file, if it is possible

serghei-rubailo commented:
Hi,

1. Changed max_length for Volunteer phone field to 8 numbers
2. Added regex to accept only digits for Volunteer phone field
3. Added /volunteer/filters/{pages}/{per_page} to swagger.
It's a pity, but I've never worked with Swagger, so if I've added new row not correctly, then could you please help me with it.

Thank you!

alexciornii commented on backend/utils/search.py:
```diff
@@ -0,0 +1,26 @@
+
+
+def model_keywords_search(model, search_fields, search_words_list, search_result=None):
+    try:
```
Please use [NumPy style docstring](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html)

serghei-rubailo replied:
I've added function description, but I'm not sure about each parameter and return description.
Because in return there is QuerySet type, but I can't set this type for return in docstring.
Could you please advice which type is right for QuerySet return type

alexciornii replied:
I think this looks good. Thanks.

serghei-rubailo commented on backend/models/volunteer_model.py:
```diff
@@ -67,4 +67,4 @@ class Volunteer(Document):
     sent_photo = BooleanField(default=False)
     aggreed_terms = BooleanField(default=False)
-    april1 = BooleanField(default=False)
+    april1 = BooleanField(default=False)
```
done

serghei-rubailo commented:
Hi!

1. Added new line in Volunteer model
2. Added function model_keywords_search description
